cider/exec/module/CiderRuntimeModule.cpp

```diff
@@ -317,172 +418,87 @@ void CiderRuntimeModule::processNextBatch(const CiderBatch& in_batch) {
   }
 }
 
-#define INDEX_SIZE 8
-#define EIGHT_BYTES 8
 void CiderRuntimeModule::fetchNonBlockingResults(int32_t start_row,
                                                  int64_t* flattened_out,
-                                                 CiderBatch& outBatch,
-                                                 std::vector<size_t> column_size) {
+                                                 CiderBatch& outBatch) {
   // follow IR write logic, it writes out results as flattened row
   // it always use 8 bytes(an int64 or double or 64bit pointer) to represent one valid
   // value. The first 8 bytes is row index, which means row id of the original input.
   // following values are column value of this row, numeric values use 64 bits(8 bytes).
   // For string type, it uses 2 64-bits value, the first is pointer, the latter is length.
-  CHECK_EQ(outBatch.column_num(), column_size.size());
   size_t col_num = outBatch.column_num();
   int64_t row_num = outBatch.row_num();
 
-  // step is row size in 8 bytes, equal to query_mem_desc->getRowSize() / 8
-  size_t step = ciderCompilationResult_->impl_->query_mem_desc_->getRowSize() / 8;
-  std::vector<size_t> offsets(col_num, 0);
-  for (int i = 0; i < col_num; i++) {
-    offsets[i] = ciderCompilationResult_->impl_->query_mem_desc_->getColOffInBytes(i) /
-                 EIGHT_BYTES;
-  }
-
   auto schema = ciderCompilationResult_->getOutputCiderTableSchema();
   if (!outBatch.schema()) {
     outBatch.set_schema(std::make_shared<CiderTableSchema>(schema));
   }
+
+  const auto& query_mem_desc = ciderCompilationResult_->impl_->query_mem_desc_;
+  const auto col_slot_context = query_mem_desc->getColSlotContext();
```

Comment:
can you point out what kind of case will original offset not work well?

Reply:
the getColOffInBytes is not correct when column has more than one slot, e.g. non-encoding sting, so the offset will always be 1,2,3,4,5....